src/gt4py/next/program_processors/runners/dace_iterator/workflow.py

```diff
@@ -164,17 +199,41 @@ class Meta:
 
 
 def convert_args(
-    inp: stages.CompiledProgram,
+    inp: CompiledDaceProgram,
     device: core_defs.DeviceType = core_defs.DeviceType.CPU,
     use_field_canonical_representation: bool = False,
 ) -> stages.CompiledProgram:
-    sdfg_program = cast(CompiledSDFG, inp)
-    on_gpu = True if device == core_defs.DeviceType.CUDA else False
+    sdfg_program = inp.sdfg_program
     sdfg = sdfg_program.sdfg
+    on_gpu = True if device == core_defs.DeviceType.CUDA else False
 
     def decorated_program(
         *args, offset_provider: dict[str, common.Connectivity | common.Dimension]
     ):
+        if sdfg_program._lastargs:
+            # The scalar arguments should be replaced with the actual value; for field arguments,
+            # the data pointer should remain the same otherwise fast-call cannot be used and
+            # the args list needs to be reconstructed.
+            use_fast_call = True
+            scalar_args_table: dict[str, ctypes._SimpleCData] = {}
+            for pos, (param, arg) in enumerate(zip(sdfg.arg_names, args), start=1):
+                if isinstance(arg, common.Field):
```

Comment:
Here you also have to check the underlying SDFG type in the signature.

Reply:
I see the point, but how likely it is that we pass the same underlying buffer but with different numpy data type? If we have to account for all cases, it is better to not use fastcall.

Reply:
As far as I know/see NumPy uses something like pool allocation, as it might reuse memory of arrays that have gone out of scope.
However, you are right it is quite far fetched.

But it should be still better to use fast call, since we are in a bit different setting.